cpp/src/copying/gather.cuh

```diff
@@ -0,0 +1,331 @@
+		  cudaStream_t stream) {
+    column_type const *source_data{source_column.data<column_type>()};
+    column_type *destination_data{destination_column.data<column_type>()};
+
+    size_type const num_destination_rows{destination_column.size()};
+
+    if (ignore_out_of_bounds) {
+      thrust::gather_if(rmm::exec_policy(stream)->on(stream), gather_map,
+			gather_map + num_destination_rows, gather_map,
+			source_data, destination_data,
+			bounds_checker<decltype(*gather_map)>{0, source_column.size()});
+    } else {
+      thrust::gather(rmm::exec_policy(stream)->on(stream), gather_map,
+		     gather_map+num_destination_rows, source_data,
+		     destination_data);
+    }
+
+    CHECK_STREAM(stream);
+  }
+
+  template <typename column_type, typename iterator_type,
+    std::enable_if_t<not std::is_integral<column_type>::value and
+                     not std::is_floating_point<column_type>::value>* = nullptr>
+  void operator()(column_view const& source_column,
+		  iterator_type gather_map,
+		  mutable_column_view& destination_column,
+		  bool ignore_out_of_bounds,
+		  util::cuda::scoped_stream stream) {
+    CUDF_FAIL("Column type must be numeric");
+  }
+
+};
+
+/**
+ * @brief Specifies the behavior of index_converter
+ */
+enum index_conversion {
+    NEGATIVE_TO_POSITIVE = 0,
+    NONE
+};
+
+/**---------------------------------------------------------------------------*
+* @brief Function object for applying a transformation on the gathermap
+* that converts negative indices to positive indices
+*
+* A negative index `i` is transformed to `i + size`, where `size` is
+* the number of elements in the column being gathered from.
+* Allowable values for the index `i` are in the range `[-size, size)`.
+* Thus, when gathering from a column of size `10`, the index `-1`
+* is transformed to `9` (i.e., the last element), `-2` is transformed
+* to `8` (the second-to-last element) and so on.
+* Positive indices are unchanged by this transformation.
+*---------------------------------------------------------------------------**/
+template <typename map_type, index_conversion ic=index_conversion::NONE>
+struct index_converter : public thrust::unary_function<map_type,map_type>{};
+
+
+template <typename map_type>
+struct index_converter<map_type, index_conversion::NEGATIVE_TO_POSITIVE>
+{
+  index_converter(size_type n_rows)
+  : n_rows(n_rows) {}
+
+  __device__
+  map_type operator()(map_type in) const
+  {
+    return ((in % n_rows) + n_rows) % n_rows;
+  }
+  size_type n_rows;
+};
+
+template <typename map_type>
+struct index_converter<map_type, index_conversion::NONE>
+{
+  index_converter(size_type n_rows)
+  : n_rows(n_rows) {}
+
+  __device__
+  map_type operator()(map_type in) const
+  {
+    return in;
+  }
+  size_type n_rows;
+};
+
+
+template <typename iterator_type>
+void gather(table_view const& source_table, iterator_type gather_map,
+	    mutable_table_view& destination_table, bool check_bounds = false,
```

Comment:
The gather cannot be in place. The output must be returned.

Reply:
Does this mean just the trivial change of allocating `destination_table` inside `gather` and returning it?

Reply:
Well, it should also go away from using `allocate_like` and instead create the output table one column at a time.

Reply:
Done